gh-125828 : Fix 'get_value' missing on [Bounded]Semaphore on multiprocessing MacOSX

YvesDup opened the pull request:
This proposal is a workaround to fix absence of 'sem_getvalue' C function in the Semaphore MacOSX implementation.

Alls unit tests succeed except on `test.test_concurrent_futures.test_init` relative to the Resource Tracker manager.


<!-- gh-issue-number: gh-125828 -->
* Issue: gh-125828
<!-- /gh-issue-number -->

YvesDup committed:
Initial commit

picnixz commented on Lib/multiprocessing/synchronize.py:
```diff
@@ -123,21 +123,89 @@ def _make_name():
+    # Specific MacOSX Semaphore
+    #
+
+    class _MacOSXSemaphore(SemLock):
```
Wouldn't it be better to handle the SemLock class differently at the C level?

YvesDup replied:
> Wouldn't it be better to handle the SemLock class differently at the C level?

As I mentioned in the problem, I'm waiting for feedback (from the Mac OS team ?) before going any further. 

A C implementation is an option, even if it seems more complicated than Python. 
That said, I'm available to go deep into the SemLock C code

picnixz replied:
Actually, since the SemLock class is implemented in C, it's preferrable to first patch it there IMO. But let's CC some people I know they are on macOS: @ned-deily @ronaldoussoren @freakboy3742.

freakboy3742 replied:
I can't profess any particular expertise in this area. However, given the underlying implementation is written in C, I agree that it makes some sense for the implementation of this workaround to also be in C (rather than have an underlying implementation that returns NotImplemented, wrapped by a layer that overrides that.

YvesDup replied:
> rather than have an underlying implementation that returns NotImplemented, wrapped by a layer that overrides that.

The main idea is to listen to the `acquire` and `release` methods to update a shared counter, and to return the counter value when the `get_value` method is invoked.

I am going to open a new PR with a workaround written in C.

picnixz commented on Lib/multiprocessing/synchronize.py:
```diff
@@ -123,21 +123,89 @@ def _make_name():
+        """
+
+        def __init__(self, kind, value, maxvalue, *, ctx):
+            util.debug(f"_MacOSXSemaphore:: creation of a {self.__class__.__name__}"\
+                        f"with '{value = }'")
```
```suggestion
            util.debug(f"_MacOSXSemaphore:: creation of a {self.__class__.__name__}"\
                       f"with {value=!r}")
```

picnixz commented on Lib/multiprocessing/synchronize.py:
```diff
@@ -123,21 +123,89 @@ def _make_name():
+        def _acquire(self, *args, **kwargs) -> bool:
+            if self._semlock.acquire(*args, **kwargs):
+                with self._count:
+                    util.debug(f"_MacOSXSemaphore: acquire {repr(self)}")
```
```suggestion
                    util.debug(f"_MacOSXSemaphore: acquire {self!r}")
```

picnixz commented on Lib/multiprocessing/synchronize.py:
```diff
@@ -123,21 +123,89 @@ def _make_name():
+            with self._count:
+                self._count.value += 1
+                self._semlock.release()
+                util.debug(f"_MacOSXSemaphore: release {repr(self)}")
```
```suggestion
                util.debug(f"_MacOSXSemaphore: release {self!r}")
```

picnixz reviewed (commented):
Is there an alternative way to do it without a temporary class like this? (I don't have the answer but I feel it overcomplicates the code base)

ZeroIntensity commented:
This needs a NEWS entry :)

ronaldoussoren commented:
Is ``get_value`` part of the public API for these classes? The name suggests it is, but the method is not document in the reference docs and does not have a docsstring.   

I haven't looked with enough detail at the code to be certain,  and am not a multiprocessing expert, but I the code doesn't seem correct.  In particular, how does the implementation handle blocking callers of acquire when the count is 0 but not when it is larger than 0?

YvesDup commented:
> is `get_value` part of the public API for these classes? The name suggests it is, but the method is not document in the reference docs and does not have a docsstring.

Yes, it is and I confirm this is not documented. This public method is rarely used to get a value of semaphore, except in unit tests. I should open an issue about this case.

> In particular, how does the implementation handle blocking callers of acquire when the count is 0 but not when it is larger than 0?

As I explain it [above](https://github.com/python/cpython/pull/125829#discussion_r1827675808), this class listen specially to `acquire` and  `release` to update shared counter then call respectively  `acquire` and `release`methods of `_semlock` class. When `get_value` is invoked, the class returns the counter value, and never calls the `semlock._get_value` method (which raises NotImplementedError). 

In this way, the behaviour of semaphores is always done at a low level in the implementation of the `semlock`class in the semaphore module (at C level).

YvesDup committed:
Merge branch 'python:main' into semaphore-macosx-multiprocessing

YvesDup committed:
Merge branch 'main' into semaphore-macosx-multiprocessing

YvesDup committed:
Merge branch 'python:main' into semaphore-macosx-multiprocessing

YvesDup committed:
Merge branch 'main' into semaphore-macosx-multiprocessing

YvesDup committed:
Merge branch 'main' into semaphore-macosx-multiprocessing

YvesDup committed:
Merge branch 'python:main' into semaphore-macosx-multiprocessing

YvesDup committed:
Refactor code

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

YvesDup committed:
Merge branch 'python:main' into semaphore-macosx-multiprocessing

YvesDup committed:
Semaphore macosx multiprocessing fix (#6)

* Refactor code
* fix error on get_value, and nits

YvesDup committed:
last modifications as:
+ Add `pending_acquires` counter
+ Update comments
+ Fix nits

YvesDup committed:
Merge branch 'main' into semaphore-macosx-multiprocessing

YvesDup committed:
Merge branch 'main' into semaphore-macosx-multiprocessing

YvesDup committed:
Update news file

Clarified the fix for the unimplemented get_value method and specified the class hierarchy for better understanding.

YvesDup committed:
Update following the execution of “regen-all”

YvesDup committed:
Fix bad class name

YvesDup committed:
Merge branch 'main' into semaphore-macosx-multiprocessing

YvesDup committed:
revert update of ./Programs/test_frozenmain.h

YvesDup committed:
Merge branch 'main' into semaphore-macosx-multiprocessing

YvesDup committed:
Merge branch 'main' into semaphore-macosx-multiprocessing

YvesDup committed:
Merge branch 'main' into semaphore-macosx-multiprocessing
